thor/main.py

```diff
@@ -137,42 +137,121 @@ def clusterVelocity(
         If clusters are found, will return a list of numpy arrays containing the 
         observation IDs for each cluster. If no clusters are found, will return np.NaN.
     """ 
+    logger.debug(f"cluster: vx={vx} vy={vy} n_obs={len(obs_ids)}")
     xx = x - vx * dt
     yy = y - vy * dt
+
+    X = np.stack((xx, yy), 1)
+
+    clusters = find_clusters(X, eps, min_samples, alg=alg)
+    clusters = filter_clusters_by_length(clusters, dt, min_samples, min_arc_length)
+
+    cluster_ids = []
+    for cluster in clusters:
+        cluster_ids.append(obs_ids[cluster])
+
+    if len(cluster_ids) == 0:
+        cluster_ids = np.NaN
+
+    return cluster_ids
+
+
+def filter_clusters_by_length(clusters, dt, min_samples, min_arc_length):
+    """
+    Filter cluster results on the conditions that they span at least
+    min_arc_length in the time dimension, and that each point in the cluster
+    is from a different dt value.
+
+    Parameters
+    -----------
+    clusters: `list of numpy.ndarray'
+
+        A list of clusters. Each cluster should be an array of indexes
+        of observations that are members of the same cluster. The indexes
+        are into the 'dt' array.
+
+    dt: `~numpy.ndarray' (N)
+        Change in time from the 0th exposure in units of MJD.
+
+    min_samples: int
+        Minimum size for a cluster to be included.
+
+    min_arc_length: float
+        Minimum arc length in units of days for a cluster to be accepted.
+
+    Returns
+    -------
+    list of numpy.ndarray
+
+        The original clusters list, filtered down.
+    """
+    filtered_clusters = []
+    for cluster in clusters:
+        dt_in_cluster = dt[cluster]
+        num_obs = len(dt_in_cluster)
+        arc_length = dt_in_cluster.max() - dt_in_cluster.min()
+        if ((num_obs == len(np.unique(dt_in_cluster)))
```

Comment:
This is a constraint from the original code, and one which surprised me a bit: we require that _all_ points in the cluster be from different `dt`s. This means that if there are any pairs of observations which are within `eps` of each other in a single exposure after applying velocity transformations, then we will reject any clusters that they end up in. 

I think this will effectively filter out any moving objects that pass in front of some other observation. Is that correct?

Reply:
Correct! It filters out observations on top of each other or within the clustering radius of each other from the same exposure time. This is definitely a feature we might want to revisit. In our initial runs on simulated data we had a lot of what we called "duplicate visit" clusters and this was a way to get around creating those. The combination of making sure each observation is at a unique time and that the minimum arc length is longer than a day drastically reduced the number of clusters to work with at almost no cost to completeness.